src/array.jl

```diff
@@ -71,9 +71,21 @@ mutable struct CuArray{T,N,M} <: AbstractGPUArray{T,N}
     else
       maxsize
     end
-    data = DataRef(pool_free, pool_alloc(M, bufsize))
-    obj = new{T,N,M}(data, maxsize, 0, dims)
-    finalizer(unsafe_free!, obj)
+
+    function _alloc_f()
+        data = DataRef(pool_free, pool_alloc(M, bufsize))
+        obj = new{T,N,M}(data, maxsize, 0, dims)
+        finalizer(unsafe_free!, obj)
+    end
+
+    name = GPUArrays.CacheAllocatorName[]
+    # Do not use caching allocator if it is not set or
+    # the buffer is not a device memory.
+    return if !(M <: DeviceMemory) || name == :none
```

Comment:
I take it you're only caching device memory because the cache doesn't differentiate between memory types and could thus return the wrong array? Would it be possible to resolve that by including the array type in the key, or do you want to be able to alias e.g. a `[1, 100]` and `[10, 10]` array?